### PR TITLE
[IMP] sale,payment: custom payment link

### PR DESCRIPTION
**Pull request by sza-odoo**
**Added custom payment link**

- Added action 'Generate a payment link' in dropdown menu
- Reference id is prefilled.
- Based on change in amount, new link is generated

Task 1938635

### CODE REVIEW

File: addons/sale/wizard/sale_generate_payment_link.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        return res
+
+    order_id = fields.Many2one("sale.order", string="Sales order", require=True)
+    amount = fields.Float(required=True)
+    link = fields.Char(string="Payment link")
+    description = fields.Char(string="Payment Ref")
+
+    def generate_payment_link(self):
+        self.ensure_one()
+        if self.order_id.state in ['draft', 'cancel']:
```
Comment (Feyensv) [inline addons/sale/wizard/sale_generate_payment_link.py]: In the domain of the action

File: addons/sale/wizard/sale_generate_payment_link.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        res['order_id'] = active_id
+        res['description'] = record.name
+        res['amount'] = record.amount_total - invoiced_amount
+        return res
+
+    order_id = fields.Many2one("sale.order", string="Sales order", require=True)
+    amount = fields.Float(required=True)
+    link = fields.Char(string="Payment link")
+    description = fields.Char(string="Payment Ref")
+
```
Comment (Feyensv) [inline addons/sale/wizard/sale_generate_payment_link.py]: Use and show currency_id on wizard as monetary widget (or make amount a monetary field)

File: addons/payment/controllers/portal.py [hunk 3]
```diff
@@ -271,7 +273,7 @@ def payment_token(self, pm_id, reference, amount, currency_id, return_url=None,
     def confirm(self, **kw):
         tx_id = int(kw.get('tx_id', 0))
         if tx_id:
-            tx = request.env['payment.transaction'].browse(tx_id)
+            tx = request.env['payment.transaction'].sudo().browse(tx_id)
```
Comment (dbo-odoo) [inline addons/payment/controllers/portal.py]: why this changes?

Comment (mba-odoo) [reply addons/payment/controllers/portal.py]: @dbo-odoo , when the payment is processed through the public user by direct link at that time it will not able to browse the transaction and due to that user was redirected to /my/home page instead of the transaction confirmation page.

**Comment by dbo-odoo** [comment]
@robodoo r+ rebase-merge plz
